docs: adds oauth plain authentication flows (#6262)
strimzi/strimzi-kafka-operator: documentation/modules/oauth/con-oauth-authentication-flow.adoc

OAUTHBEARER provides a higher level of security than PLAIN because client credentials are _never_ shared with Kafka brokers.
Consider using PLAIN only with Kafka clients that do not support OAUTHBEARER.

You configure Kafka broker listeners to use OAuth 2.0 authentication for clients connection.
If necessary, you can enable OAUTHBEARER and PLAIN on the same `oauth` listener.

.OAUTHBEARER overview

OAUTHBEARER is automatically enabled in the `oauth` listener configuration for the Kafka broker.
You can set the `enableOauthBearer` property to `true`, though this is not required.

[source,yaml,subs="attributes+"]
----
  # ...
  authentication:
    type: oauth
    # ...
    enableOauthBearer: true
----

> The OAUTHBEARER authentication mechanism must be explicitly configured.

Author: Bit vague. I think we're talking about the actual _config_ rather than enabling. It applies to both mechanisms anyway, so I've moved up into the intro. I'll see what @mstruk thinks.

> You configure Kafka broker listeners to use OAuth 2.0 authentication for connecting clients.
> If necessary, you can use the OAUTHBEARER and PLAIN mechanisms on the same `oauth` listener.
> The properties to support each mechanism must be explicitly specified in the `oauth` listener configuration.

OAUTHBEARER provides a higher level of security than PLAIN because client credentials are _never_ shared with Kafka brokers.
Consider using PLAIN only with Kafka clients that do not support OAUTHBEARER.

You configure Kafka broker listeners to use OAuth 2.0 authentication for connecting clients.
If necessary, you can use the OAUTHBEARER and PLAIN mechanisms on the same `oauth` listener.
The properties to support each mechanism must be explicitly specified in the `oauth` listener configuration. 

.OAUTHBEARER overview

OAUTHBEARER is automatically enabled in the `oauth` listener configuration for the Kafka broker.
You can set the `enableOauthBearer` property to `true`, though this is not required.

[source,yaml,subs="attributes+"]
----
  # ...
  authentication:
    type: oauth
    # ...
    enableOauthBearer: true
----

Many Kafka client tools use libraries that provide basic support for OAUTHBEARER at the protocol level.
To support application development, Strimzi provides an _OAuth callback handler_ for the upstream Kafka Client Java libraries (but not for other libraries).

> Therefore, you do not need to write your own callback handlers for such clients.

Reviewer: Due to the last part in backets:
> (but not for other libraries)

One might interpret 'such clients' to refer to these other libraries?

```suggestion
Therefore, you do not need to write your own callback handlers.
```

* A long-lived access token, obtained manually at configuration time

For both methods, the client must provide the PLAIN `username` and `password` properties to pass credentials to the Kafka broker.
The client uses these properties to pass a client ID and secret or username and access token.
A client ID and secret are used to obtain an access token.

For the Kafka broker to interpret a password as an access token, the following configuration is required:

* The broker listener configuration must have a token endpoint (`tokenEndpointUri`) configured
* The password requires an `$accessToken:` prefix followed by the value of the access token

If a token endpoint is not configured on the broker listener, the Kafka broker interprets the password as a raw access token.
In this case, the password should not include a prefix.

If the `password` is set as the access token, the `username` must be set to the same principal name that the Kafka broker obtains from the access token.
The process depends on how you configure username extraction using `userNameClaim`, `fallbackUserNameClaim`, `fallbackUsernamePrefix`, or `userInfoEndpointUri`.

> It also depends on your authorization server; in particular, how it maps client IDs to account names.

Reviewer 2: 'It' should be specified. Sentence is vague

Author: Changed to:

> You can specify username extraction options in your listener using the `userNameClaim`, `fallbackUserNameClaim`, `fallbackUsernamePrefix`, and `userInfoEndpointUri` properties.
> The username extraction process also depends on your authorization server; in particular, how it maps client IDs to account names.

* A long-lived access token, obtained manually at configuration time

> For both methods, the client must provide a `username` and `password` to pass credentials to the Kafka broker.

Reviewer: ```suggestion
For both methods, the client must provide `username` and `password` PLAIN parameters to pass credentials to the Kafka broker.
```

> If the password is prefixed with `$accessToken:` followed by the value of an access token, the Kafka broker will interpret the password as the access token.

Reviewer: ```suggestion
When listener is configured with `tokenEndpointUri`, if the password is prefixed with `$accessToken:` followed by the value of an access token, the Kafka broker will interpret the password as the access token.
```

Author: Tried to break a very long sentence up a bit.

> For both methods, the client must provide the PLAIN properties `username` and `password` to pass credentials to the Kafka broker.
> For the Kafka broker to interpret the password as an access token, it requires an `$accessToken:` prefix followed by the value of the access token. The listener configuration must have a token endpoint (`tokenEndpointUri`) configured for authentication. 
> Otherwise, the Kafka broker will interpret the `username` as the client ID and the `password` as the client secret.
>

Otherwise, the Kafka broker will interpret the `username` as the client ID and the `password` as the client secret.

> 

Reviewer: ```suggestion
When listener has no `tokenEndpointUri` configured, the `password` parameter is never interpreted as a `secret` since it can't be exchanged for an access token without a `tokenEndpoinyUri`. In this case `secret` is always interpreted as an unprefixed `access token`. 
```

Author: Not sure I understand what we're saying here. Is this to replace the preceding sentence, which seems to contradict it?

What we have so far:

> For both methods, the client must provide the PLAIN `username` and `password` properties to pass credentials to the Kafka broker.
> For the Kafka broker to interpret a password as an access token, the following is required:
> 
> * The listener configuration must have a token endpoint (`tokenEndpointUri`) configured 
> * The password requires an `$accessToken:` prefix followed by the value of the access token
> 
> Otherwise, the Kafka broker will interpret the `username` as the client ID and the `password` as the client secret.

Reviewer: > If a token endpoint is not configured on the broker listener, ...

Doesn't that sound strange, to first make a categorical claim in the previous paragraph that two conditions are required to use access token for authentication, but in the next sentence add another case (negating that the two conditions are required) where we say that if the first condition is not fulfilled then it's also possible to use access token. Maybe someone else can chime in (@scholzj ?). I would prefer to start from server config being with tokenEndpointUri or without tokenEndpointUri and depending on that explain the options for credentials, rather than trying to explain how to pass access token which is the angle from which the explanation goes here. But maybe that's just me.
 
> the password should not include

Maybe use `the password parameter` or `the password property` to make it clear we have _that_ password in mind.

Author: Thnaks, @mstruk. Let's rearrange a little:

> For both methods, the client must provide the PLAIN `username` and `password` properties to pass credentials to the Kafka broker. The client uses these properties to pass a client ID and secret or username and access token.
> 
> Client IDs and secrets are used to obtain access tokens.
> 
> Access tokens are passed as `password` property values. 
> You pass the access token with or without an `$accessToken:` prefix. 
> 
> * If you configure a token endpoint (`tokenEndpointUri`) in the listener configuration, you need the prefix.
> * If you don't configure a token endpoint (`tokenEndpointUri`) in the listener configuration, you don't need the prefix. The Kafka broker interprets the password as a raw access token.

Reviewer: Yeah I think that's better.
